### DISCUSSION TITLE
Better Gray to Black and White (BW) conversion

### DISCUSSION THREAD

**Discussion by duaneg3** [python-pillow/Pillow#7850] [Q&A] [answered]
I want to implement an adaptive thresholding rather than a simple thresholding conversion from Gray to BW.  The adaptive thresholding code is as follows:
```python
    imgx = my_img.filter(ImageFilter.GaussianBlur(3)) 
    xlist = collections.deque(imgx.getdata())
    ilist = collections.deque(my_img.getdata())
    olist= []
    while(len(ilist) > 0):
        ia = ilist.popleft()
        ix = xlist.popleft()
        if ia < 127 or ia < (ix-4):  olist.append(0)
        else: olist.append(255)
    img1 = Image.new("1", my_img.size)
    img1.putdata(olist)
```
However it runs fairly slowly compared simple thresholding using point.
```python
    img1 = img1a.point(lambda p: p > 126 and 255)  
```
Is there a more performant way to implement adaptive thresholding?
The ImageMath module would seem appropriate.  However, I do not see a way to compare pixel values and pick two different values a result of this comparison.

**Comment by Yay295** [comment]
I don't think it will make a huge difference, but this should be a little bit faster:
```python
imgx = my_img.filter(ImageFilter.GaussianBlur(3))
olist = []
for ia,ix in zip(my_img.getdata(),imgx.getdata()):
    if ia < 127 or ia < (ix-4): olist.append(0)
    else: olist.append(255)
img1 = Image.frombytes("1",my_img.size)
img1.putdata(olist)
```

**Comment by Yay295** [comment]
I think NumPy would be your best option for this. I have not tested this.

```python
imgx = my_img.filter(ImageFilter.GaussianBlur(3))
npai = np.asarray(my_img)
npax = np.asarray(imgx)
data = np.where(npai < 127 or npai < (npax-4), 0, 255)
img1 = Image.fromarray(data)
```

https://numpy.org/doc/stable/reference/generated/numpy.where.html

**Reply by duaneg3** [reply]
Thanks!  Unfortunately, they all run in about the same time.
NumPy might be a little faster, but not very noticeable.

**Comment by radarhere** [comment]
It could be helpful to know some more specifics to your situation. Are you able to upload a copy of an image that you would pass to the script? Or at least let us know its mode and dimensions?

**Reply by duaneg3** [reply]
My use case is a large multiple page PDF that came from a scanner with gray scale images of the pages.  I want to switch to blank and white images. 
Sometimes the images contain gray text.  Simple thresholding does not work on gray text.  Adaptive thresholding working well.  The "imgx" calculation is quick.  The delay is in the rest of the calculation.  A full page of text is reasonably slower with the adaptive thresholding than the simple thresholding with point()  (point is very performant).
An image of a full page of black and white text shows the performance issue quite well.  Adding in a little gray text shows the need for adaptive thresholding.
I should be able to build a test case image.

Thanks for your responses!!

**Comment by duaneg3** [comment]
Not sure how it worked before:

L
L
dimen= 2 len 14231330
dimen= 2 len 14231330
Traceback (most recent call last):
data = np.where(npai < 127 or npai < (npax-4), 0, 255)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

"L" is a gray image .. not sure why numpy thinks it has two dimensions.
Not sure why it worked before ..

 ```python
       print(my_img.mode)
        print(imgx.mode)
        npai = np.asarray(my_img)
        npax = np.asarray(imgx)
        print("dimen=",npai.ndim,"len",npai.size)
        print("dimen=",npax.ndim,"len",npax.size)
        data = np.where(npai < 127 or npai < (npax-4), 0, 255)
        img1 = Image.fromarray(data)
```

**Comment by duaneg3** [comment]
Here is a small example:

```python
#! /usr/bin/python3

from PIL import Image,ImageFile
from PIL import ImageFilter
import PIL
import numpy as np

my_img = Image.open("page_scan.jpg")
imgx = my_img.filter(ImageFilter.GaussianBlur(3)) 
print("img_mode:",my_img.mode,"imx_mode:",imgx.mode)

if 1==2:
    img1 = my_img.point(lambda p: p > 126 and 255)  

elif 1==1:
    print("converting page")
    olist= []
    for ia,ix in zip(my_img.getdata(),imgx.getdata()):
        if (ia < 127) or (ia < (ix-4)): olist.append(0)
        else: olist.append(255)
    img1 = Image.new("1", my_img.size)
    img1.putdata(olist)
    print("done converting page")

fname = "test_done.jpg"
img1.save(fname)
print("wrote:",fname);

```

![page_scan](https://github.com/python-pillow/Pillow/assets/91136015/1c35051e-f5fa-4ef6-b1d7-789a5bff5e75)

thanks in advance for any help!!

**Reply by radarhere** [reply]
Thanks. I find that

```python
olist = [0 if (ia < 127) or (ia < (ix-4)) else 255 for ia,ix in zip(my_img.getdata(),imgx.getdata())]
img1 = Image.new("1", my_img.size)
img1.putdata(olist)
```
is slightly faster.

**Comment by duaneg3** [comment, marked as answer]
I extended the ImageMath to have abw() and bw() functions.
(the library is quite nice ...)
now runs really fast ...

```python
elif 1==2:
    img1 = ImageMath.eval("bw(a, 127)", a=my_img)
elif 1==1:
    img1 = ImageMath.eval("abw(a, b, 4, 127)", a=my_img, b=imgx)
```

**Reply by radarhere** [reply]
This has been continued in #7857
